gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeBuilder.java

```diff
@@ -72,6 +86,61 @@ public QueryNode build() {
     return executionTree;
   }
 
+  private QueryNode buildExecutionTreeForSameSourceFilterAndSelection(String source) {
+    if (source.equals("QS")) {
+      return buildExecutionTreeForQsFilterAndSelection(source);
+    } else if(source.equals("EDS")) {
+      return buildExecutionTreeForEdsFilterAndSelection(source);
+    } else {
+      throw new UnsupportedOperationException("Unknown Entities data source. No fetcher for this source.");
+    }
+  }
+
+  private QueryNode buildExecutionTreeForQsFilterAndSelection(String source) {
+    int selectionLimit = executionContext.getEntitiesRequest().getLimit();
+    int selectionOffset = executionContext.getEntitiesRequest().getOffset();
+
+    // query-service/Pinot does not support offset when group by is specified. Since we will be
+    // grouping by at least the entity id, we will compute the non zero pagination ourselves. This
+    // means that we need to request for offset + limit rows so that we can paginate appropriately.
+    // Pinot will do the ordering for us.
```

Comment:
Can we push this down into query service? It's better to make GW not worry about underlying OLAP store.

Reply:
There's a case to be made for doing this.